internal/services/approleassignments/app_role_assignment_resource.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/manicminer/hamilton/msgraph"
+	"github.com/manicminer/hamilton/odata"
+
+	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
+	"github.com/hashicorp/terraform-provider-azuread/internal/services/approleassignments/parse"
+	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
+	"github.com/hashicorp/terraform-provider-azuread/internal/utils"
+	"github.com/hashicorp/terraform-provider-azuread/internal/validate"
+)
+
+func appRoleAssignmentResource() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: appRoleAssignmentResourceCreate,
+		ReadContext:   appRoleAssignmentResourceRead,
+		DeleteContext: appRoleAssignmentResourceDelete,
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(5 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Delete: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Importer: tf.ValidateResourceIDPriorToImport(func(id string) error {
+			_, err := parse.AppRoleAssignmentID(id)
+			return err
+		}),
+
+		Schema: map[string]*schema.Schema{
+			"app_role_id": {
+				Description:      "The ID of the app role to be assigned",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validate.UUID,
+			},
+
+			"principal_object_id": {
+				Description:      "The object ID of the user, group or service principal to be assigned this app role",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validate.UUID,
+			},
+
+			"resource_object_id": {
+				Description:      "The object ID of the service principal representing the resource",
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validate.UUID,
+			},
+
+			"principal_display_name": {
+				Description: "The display name of the principal to which the app role is assigned",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+
+			"principal_type": {
+				Description: "The object type of the principal to which the app role is assigned",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+
+			"resource_display_name": {
+				Description: "The display name of the application representing the resource",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func appRoleAssignmentResourceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).AppRoleAssignments.AppRoleAssignedToClient
+	servicePrincipalsClient := meta.(*clients.Client).AppRoleAssignments.ServicePrincipalsClient
+
+	appRoleId := d.Get("app_role_id").(string)
+	principalId := d.Get("principal_object_id").(string)
+	resourceId := d.Get("resource_object_id").(string)
+	_, status, err := servicePrincipalsClient.Get(ctx, resourceId, odata.Query{})
+	if err != nil {
```

Comment:
I THINK THESE LINES CAN BE MERGED?

Reply:
![nodding-sign](https://user-images.githubusercontent.com/251987/134416461-7a7e6c5b-4720-43f3-b387-f88dd1f0bbb3.gif)

We should get a linter for these :D